monitoring/asset_balances_collector.go

```diff
@@ -0,0 +1,112 @@
+package monitoring
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// assetBalancesCollector is a Prometheus collector that exports the balances
+// of all taproot assets.
+type assetBalancesCollector struct {
+	collectMx sync.Mutex
+
+	cfg      *PrometheusConfig
+	registry *prometheus.Registry
+
+	balancesVec *prometheus.GaugeVec
+
+	utxosVec *prometheus.GaugeVec
+}
+
+func newAssetBalancesCollector(cfg *PrometheusConfig,
+	registry *prometheus.Registry) *assetBalancesCollector {
+
+	return &assetBalancesCollector{
+		cfg:      cfg,
+		registry: registry,
+		balancesVec: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "asset_balances",
+				Help: "Balances of all taproot assets",
+			},
+			[]string{"asset_label"},
+		),
+		utxosVec: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "utxos_assets_held",
+				Help: "Number of UTXOs used for taproot assets",
+			},
+			[]string{"utxo_label"},
+		),
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once the
+// last descriptor has been sent.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetBalancesCollector) Describe(ch chan<- *prometheus.Desc) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	a.balancesVec.Describe(ch)
+	a.utxosVec.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetBalancesCollector) Collect(ch chan<- prometheus.Metric) {
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	if a.cfg == nil {
+		log.Error("cfg is nil")
+		return
+	}
+
+	if a.cfg.AssetStore == nil {
+		log.Error("asset store is nil")
+		return
+	}
+
+	ctxdb, cancel := context.WithTimeout(context.Background(), dbTimeout)
+	defer cancel()
+
+	assets, err := a.cfg.AssetStore.FetchAllAssets(ctxdb, false, false, nil)
```

Comment:
I think this is a bad idea as the number of assets will always just go up and this query will put an extra load on the DB as well as might just timeout if takes long. Perhaps better to build a specialized query that joins the asset and managed utxo tables and returns exactly what we want. Even then I'd say it's better to do this on-demand if only for a grafana dash and not for alerting.

Reply:
`FetchAllAssets` fetches the assets owned by the daemon. It basically gives you all your balances for all your owned assets

Reply:
> I think this is a bad idea as the number of assets will always just go up and this query will put an extra load on the DB as well as might just timeout if takes long. 

So yeah if you're hoarding an enormous amount of different asset types you could run into this, but don't think it's reasonable to think of that possibility now

Reply:
Yeah if we want to maximise performance over everything we'd add something like `FetchTagsOutpointsAmountsOfAssets` but kind of seems like polluting the store interface. We are using 3 of the most important fields already returned by `FetchAllAssets` so the justification for custom store method is bit weaker

what do others think @ffranr @guggero

Reply:
Since we're passing in `false` for `includeSpent`, this will only return assets owned by the daemon that are currently not spent. So this number shouldn't increase massively and IMO it should be fine.
I agree that we should try to keep the number of queries to a reasonable count as they are rather complex in some cases and hard to maintain (and test) if they have a lot of joins.